Domain expiry notifications

AiroPi opened the pull request:
## 📋 Overview

<!-- Provide a clear summary of the purpose and scope of this pull request:-->

- Domain expiry are monitored by some other tools, but not Uptime Kuma.
- This pull request introduce domain name expiry notification and monitoring, so we don't miss the important date !

I used RDAP APIs to get the expiry date. This worked for most of the domains I tested, but I am aware that not absolutely all domains supports this for now.

Domains lookups are saved in a database, and are checked on time a day.
If the option is activated for multiple URLs on the same domain, only one lookup a day will be done, and only one notification will be sent.

## Technical questions

Currently, the table domain_expiry_info is never emptied. Maybe some sort of cleanup could be added, how can I implement this ?
Maybe "clear all entries not checked for 2 days", because it means the check isn't active anymore.

Also, I currently fetch the RDAP list at every check (https://data.iana.org/rdap/dns.json). So this is "1 time a day * number of domains monitored". Do you think some sort of cache should be added for this ?

I would like to avert user if the TLD is not supported (no RDAP in the json), but I don't really now how to achieve this.
I currently added a WARN in the console if not RDAP server is found.

<!--
Please link any GitHub issues or tasks that this pull request addresses.
Use the appropriate issue numbers or links to enable auto-closing.
-->

- Relates to #1775
- Resolves #944

## 🛠️ Type of change

<!-- Please select all options that apply -->

- [ ] 🐛 Bugfix (a non-breaking change that resolves an issue)
- [x] ✨ New feature (a non-breaking change that adds new functionality)
- [ ] ⚠️ Breaking change (a fix or feature that alters existing functionality in a way that could cause issues)
- [x] 🎨 User Interface (UI) updates
- [ ] 📄 New Documentation (addition of new documentation)
- [ ] 📄 Documentation Update (modification of existing documentation)
- [ ] 📄 Documentation Update Required (the change requires updates to related documentation)
- [ ] 🔧 Other (please specify):
  - Provide additional details here.

## 📄 Checklist

<!-- Please select all options that apply -->

- [ ] 🔍 My code adheres to the style guidelines of this project.
- [x] 🦿 I have indicated where (if any) I used an LLM for the contributions
- [ ] ✅ I ran ESLint and other code linters for modified files.
- [x] 🛠️ I have reviewed and tested my code.
- [ ] 📝 I have commented my code, especially in hard-to-understand areas (e.g., using JSDoc for methods).
- [ ] ⚠️ My changes generate no new warnings.
- [ ] 🤖 My code needed automated testing. I have added them (this is an optional task).
- [ ] 📄 Documentation updates are included (if applicable).
- [x] 🔒 I have considered potential security impacts and mitigated risks.
- [x] 🧰 Dependency updates are listed and explained.
- [x] 📚 I have read and understood the [Pull Request guidelines](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#recommended-pull-request-guideline).

## 📷 Screenshots or Visual Changes

<!--
If this pull request introduces visual changes, please provide the following details.
If not, remove this section.

Please upload the image directly here by pasting it or dragging and dropping.
Avoid using external image services as the image will be uploaded automatically.
-->

<img width="474" height="122" alt="image" src="https://github.com/user-attachments/assets/d777c4a9-2771-4481-8295-b375eb06c05b" />
Domain Name Expiry Notification option for http monitor

<img width="464" height="464" alt="image" src="https://github.com/user-attachments/assets/f2c6ae31-b67e-43ad-8b2d-b679ed27ccde" />
Domain expiry notify days settings

<img width="962" height="195" alt="image" src="https://github.com/user-attachments/assets/c6c372ec-2991-4481-84dc-f272da62c355" />
Domain expiry stat on overview

<img width="508" height="73" alt="image" src="https://github.com/user-attachments/assets/c433cce5-f352-48a7-acaa-f5f1ec7e42e9" />
Domain expiry notification

CommanderStorm commented on server/model/monitor.js:
```diff
@@ -1461,6 +1496,110 @@ class Monitor extends BeanModel {
+        let domainInfoBean = await R.findOne("domain_expiry_info", "domain = ?", [ domain ]);
+
+        let expiryDate;
+        if (domainInfoBean === null || getDaysBetween(new Date(domainInfoBean.lastCheck), new Date()) > 1) {
```
Let's move this logic to the top via inverting the check and doing an early return

I think this check is more selective.

CommanderStorm replied:
I meant pushing `notificationList` intialisation below `expiryDate === null`.

I the end likely does not matter all that much.
Feel free to ignore 😉

AiroPi replied:
Oh ok I see now, I can swap the two 👍

AiroPi replied:
I will let the order like this, otherwise the conditions would be too messy

CommanderStorm commented on server/model/monitor.js:
```diff
@@ -1461,6 +1496,110 @@ class Monitor extends BeanModel {
+        }
+
+        const daysRemaining = getDaysRemaining(new Date(), expiryDate);
+        log.debug("domain", `${domain} expires in ${daysRemaining} days`);
```
Since we also send a notification 

```suggestion
        log.warn("domain", `${domain} expires in ${daysRemaining} days`);
```

CommanderStorm reviewed (changes requested):
Seems solid, nice work.
I have noted a few things below, but nothing structural.

Sorry to split this review into two parts, I am reviewing this from a beach in Italy 🏖️

AiroPi commented:
- [ ] I may also need to add checks if the 'url.hostname' is an IP and not a domain

CommanderStorm commented on server/model/monitor.js:
```diff
@@ -405,6 +414,11 @@ class Monitor extends BeanModel {
                     // Do not do any queries/high loading things before the "bean.ping"
                     let startTime = dayjs().valueOf();
 
+                    // Early check to support upside down monitoring
+                    if (this.isEnabledDomainExpiryNotification()) {
```
lets add a code comment explaining this, it is not obvious otherwise ^^

CommanderStorm commented on db/knex_migrations/2025-09-02-0000-add-domain-expiry-notification.js:
```diff
@@ -0,0 +1,26 @@
+        .alterTable("monitor", function (table) {
+            table.boolean("domain_expiry_notification").defaultTo(0);
+        })
+        .createTable("domain_expiry_info", (table) => {
+            table.increments("id");
+            table.datetime("last_check");
+            table.text("domain").unique();
+            table.datetime("expiry");
+        })
+        .createTable("domain_expiry_notification_sent_history", (table) => {
+            table.increments("id");
+            table.text("domain");
+            table.integer("days");
+        });
```
> With `domain_expiry_notification_sent_history` we know if a notification for 'x' days before has already been sent or not.
> I don't think we can merge the two tables, because one needs to know if a notification has been sent for 'x' days before, for 'y' days before, etc... so 'domain' column can't be unique.

I still don't see how this table is nessesary.
Is when we `last_checked` not already stored in said column?

If a domain is truly expiring, I would not have trouble telling people daily.

CommanderStorm replied:
Yea sorry, I really should not have started the Review if I don't have the time to finish it (work really kicked up a few gears the last month)

I still think we can simplify this to one value:
> So we need to track if the notification for 10 days before has already been sent, same for the 5 days before, etc...
This is tracked in the domain_expiry_notification_sent_history

We actually don't. We just need to know when the last notification for a domain has been send.

Regarding your algorithm, in case Uptime Kuma is down for a day, I think the following logic is better 😉
````py
for day in expiry_notification_days:
  if last_expiry_notification_sent < day:
     continue

  send_notification()
  last_expiry_notification_sent = day
```

CommanderStorm commented on db/knex_migrations/2025-09-02-0000-add-domain-expiry-notification.js:
```diff
@@ -0,0 +1,26 @@
+exports.up = function (knex) {
+    return knex.schema
+        .alterTable("monitor", function (table) {
+            table.boolean("domain_expiry_notification").defaultTo(0);
```
I don't think this is that common. Lets add this as default=`true`

```suggestion
            table.boolean("domain_expiry_notification").defaultTo(1);
```

or if this works:

```suggestion
            table.boolean("domain_expiry_notification").defaultTo(true);
```

AiroPi commented on server/util-server.js:
```diff
@@ -626,6 +627,80 @@ const getDaysRemaining = (validFrom, validTo) => {
+async function getRdapServer(tld) {
+    let rdapList;
+    try {
+        const res = await fetch("https://data.iana.org/rdap/dns.json");
```
I wasn't able to find any official informations about rate limit / server side cache, but the request returns some cache headers, and I was able to make a great amount of request, without hitting any limit.

Considering a case when we monitor 100 different domains, that would be 100 requests a day, and not at the same time, so even if caching this request would be a must, it should cause any issues to not cache it.

AiroPi commented:
> Any progress here @AiroPi @CommanderStorm ?

Hi, on my side I continue to fix the merge conflicts when they occurs (I will probably fix the new conflicts tomorrow or this week end). I addressed all the points of the previous review and the PR is waiting for a new review / tests.

Feel free to test it/review it, that would help the maintainers as they don't have time for this PR right now unfortunately.
